File: ruby/webhook.rb
require 'sinatra'
require 'json'

post '/' do
  request.body.rewind
  payload_body = request.body.read
  puts "==============================================================="
  puts payload_body
  puts "==============================================================="
  verify_signature(payload_body)
  push = JSON.parse(payload_body)
  puts "Topic Recieved: #{push['topic']}"
end

def verify_signature(payload_body)
  secret = "secret"
  expected = request.env['HTTP_X_HUB_SIGNATURE']
  if expected.nil? || expected.empty? then
    puts "Not signed. Not calculating"
  else
    signature = 'sha1=' + OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha1'), secret, payload_body)
    puts "Expected  : #{expected}"
    puts "Calculated: #{signature}"
    if Rack::Utils.secure_compare(signature, expected) then
      puts "   Match"
    else
      puts "   MISMATCH!!!!!!!"
      return halt 500, "Signatures didn't match!"
    end
  end
end

File: php/webhook.php
<?php

$stdout = fopen('php://stdout', 'w');
function readInput(){
  $inputContent = '';
  $input = fopen('php://input' , 'rb');
  while (!feof($input)) {
      $inputContent .= fread($input, 4096);
  }
  fclose($input);
  return $inputContent;
}

function printOutput($msg){
  global $stdout;
  fwrite($stdout, $msg . "\n");
}

function verify_signature($payload_body){
  $ALGO = 'sha1';
  $SECRET = "secret";
  $expected = $_SERVER['HTTP_X_HUB_SIGNATURE'];
  if(empty($expected)){
    printOutput("Not signed. Not calculating");
  }
  else{
    $calculated = "sha1=" . hash_hmac($ALGO, $payload_body, $SECRET, false);
    printOutput("Expected  : " . $expected);
    printOutput("Calculated: " . $calculated);
    printOutput("Match?    : " . ($expected === $calculated ? "true" : "false"));
  }
}

$DATA = readInput();
printOutput("===============================================================");
printOutput($DATA);
printOutput("===============================================================");
verify_signature($DATA);
$push = json_decode($DATA);
printOutput("Topic Recieved: " . $push->topic);
